cps/kobo.py

```diff
@@ -392,6 +394,222 @@ def get_metadata(book):
+
+    if not shelf_lib.check_shelf_edit_permissions(shelf):
+        abort(401, description="User is unauthaurized to edit shelf.")
+
+    items_unknown_to_calibre = []
+    for item in items:
+        if item["Type"] != "ProductRevisionTagItem":
+            items_unknown_to_calibre.append(item)
+            continue
+
+        book = db.session.query(db.Books).filter(db.Books.uuid == item["RevisionId"]).one_or_none()
+        if not book:
+            items_unknown_to_calibre.append(item)
+            continue
+
+        shelf.books.filter(ub.BookShelf.book_id == book.id).delete()
+    ub.session.commit()
+
+    if items_unknown_to_calibre:
+        log.debug("Received request to remove an unknown book to a collecition. Silently ignoring item.")
+
+    return make_response('', 200)
+
+
+# Add new, changed, or deleted shelves to the sync_results.
+# Note: Public shelves that aren't owned by the user aren't supported.
+def sync_shelves(sync_token, sync_results):
+    new_tags_last_modified = sync_token.tags_last_modified
+
+    for shelf in ub.session.query(ub.ShelfArchive).filter(func.datetime(ub.ShelfArchive.last_modified) > sync_token.tags_last_modified, ub.ShelfArchive.user_id == current_user.id):
+        new_tags_last_modified = max(shelf.last_modified, new_tags_last_modified)
+
+        sync_results.append({
+            "DeletedTag": {
+                "Tag": {
+                    "Id": shelf.uuid,
+                    "LastModified": convert_to_kobo_timestamp_string(shelf.last_modified)
+                }
+            }
+        })
+
+    for shelf in ub.session.query(ub.Shelf).filter(func.datetime(ub.Shelf.last_modified) > sync_token.tags_last_modified, ub.Shelf.user_id == current_user.id):
+        if not shelf_lib.check_shelf_view_permissions(shelf):
+            continue
+
+        new_tags_last_modified = max(shelf.last_modified, new_tags_last_modified)
+
+        tag = create_kobo_tag(shelf)
+        if not tag:
+            continue
+
+        if shelf.created > sync_token.tags_last_modified:
+            sync_results.append({
+                "NewTag": tag
+            })
+        else:
+            sync_results.append({
+                "ChangedTag": tag
+            })
+    sync_token.tags_last_modified = new_tags_last_modified
+    ub.session.commit()
+
+
+# Creates a Kobo "Tag" object from a ub.Shelf object
+def create_kobo_tag(shelf):
+    tag = {
+        "Created": convert_to_kobo_timestamp_string(shelf.created),
+        "Id": shelf.uuid,
+        "Items": [],
+        "LastModified": convert_to_kobo_timestamp_string(shelf.last_modified),
+        "Name": shelf.name,
+        "Type": "UserTag"
+    }
+    for book_shelf in shelf.books:
+        book = db.session.query(db.Books).filter(db.Books.id == book_shelf.book_id).one_or_none()
+        if not book:
+            log.info(u"Book (id: %s) in BookShelf (id: %s) not found in book database",  book_shelf.book_id, shelf.id)
+            return None
```

Comment:
This code returns on the first not found book in a shelf, shouldn't it be `continue` instead, to get other books to be added in a shelf?

Reply:
Fixed in 06c15a7